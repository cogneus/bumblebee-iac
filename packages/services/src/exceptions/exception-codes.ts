import { ExceptionCodes } from 'bumblebee-common'

export const TemplateExceptionCodes = {
  ...ExceptionCodes,
  TemplateNotFound: 'TemplateNotFound',
}
