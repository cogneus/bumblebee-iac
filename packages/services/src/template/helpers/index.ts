export * from './keys.helper'
export * from './validate-template.helper'
export * from './generate-id'
export * from './filter-input.class'
