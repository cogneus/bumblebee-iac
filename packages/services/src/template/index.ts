export * from './add'
export * from './get'
export * from './query'
export * from './remove'
export * from './update'
export * from './versions'
