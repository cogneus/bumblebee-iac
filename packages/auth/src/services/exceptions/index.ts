import { StatusCodes, AppException } from 'bumblebee-common'
import { AuthExceptionCodes } from './exception-codes'

export const KeyNotFound = (key: string) =>
  new AppException({
    code: AuthExceptionCodes.KeyNotFound,
    status: StatusCodes.Forbidden,
    message: `Key with name '${key}' not found`,
    source: KeyNotFound,
  })
