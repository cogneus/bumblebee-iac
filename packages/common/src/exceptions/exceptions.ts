import { AppException } from './app-exception.class'
import { ExceptionCodes } from './exception-codes'
import { StatusCodes } from './status-codes'

export const InvalidToken = () =>
  new AppException({
    code: ExceptionCodes.InvalidToken,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidToken,
  })

export const InvalidAuthorizartion = () =>
  new AppException({
    code: ExceptionCodes.InvalidAuthorizartion,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidAuthorizartion,
  })

export const EmptyRequest = () =>
  new AppException({
    code: ExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Request data is missing',
    source: EmptyRequest,
  })

export const InvalidService = (service: string) =>
  new AppException({
    code: ExceptionCodes.InvalidService,
    status: StatusCodes.Error,
    message: `Invalid service. ${service}`,
    source: InvalidService,
  })

export const InvalidRequest = (innerMessage: string) =>
  new AppException({
    code: ExceptionCodes.InvalidRequest,
    status: StatusCodes.BadRequest,
    message: `Request data is invalid. ${innerMessage}`,
    source: InvalidRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })

export const EnvVarNotFound = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.EnvVarNotFound,
    status: StatusCodes.Error,
    message: `The following environment variables not found ${keys.join()}`,
    source: EnvVarNotFound,
  })

export const InvalidModeSource = (resourcePrefix: string, eventSourceARN: string | undefined) =>
  new AppException({
    code: ExceptionCodes.InvalidModeSource,
    status: StatusCodes.Error,
    message: `Cannot derive stage from event source. eventSourceARN: ${eventSourceARN} RESOURCE_PREFIX: ${resourcePrefix}`,
    source: InvalidModeSource,
  })
