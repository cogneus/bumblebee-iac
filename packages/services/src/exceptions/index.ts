import { StatusCodes, AppException } from 'bumblebee-common'
import { TemplateExceptionCodes } from './exception-codes'

export const TemplateNotFound = (id: string) =>
  new AppException({
    code: TemplateExceptionCodes.TemplateNotFound,
    status: StatusCodes.NotFound,
    message: `Template with id '${id}' not found`,
    source: TemplateNotFound,
  })

export const EmptyRequest = () =>
  new AppException({
    code: TemplateExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Template request data is missing',
    source: EmptyRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: TemplateExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })
