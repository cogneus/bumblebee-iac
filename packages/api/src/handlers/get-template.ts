import { apiGatewayWrapper, warmupWrapper, standardEnvVars, Token } from 'bumblebee-common'
import { TemplateVersion, getTemplate } from 'bumblebee-services'
import { checkAPIEnvironmentVars, checkReadPermissions, getClientOptions } from '../helpers'
import { TemplatesParams } from '../types'

checkAPIEnvironmentVars()

export const main = warmupWrapper(
  apiGatewayWrapper<any, TemplatesParams, Token, TemplateVersion>({
    handler: ({ params: { id, version }, serviceContext }) =>
      getTemplate(serviceContext, id!, version),
    ...standardEnvVars(),
    serviceName: 'template-api',
    logName: 'get-template',
    getClientOptions,
    authCheck: checkReadPermissions,
  })
)
