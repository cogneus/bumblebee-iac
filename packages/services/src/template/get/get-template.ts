import { IDBClient, ServiceContext, ensureClient } from 'bumblebee-common'
import { TemplateVersion, DBTemplateItem } from '../../types'
import { removeKeyInfo } from '../helpers'
import { MissingAttributes, TemplateNotFound } from '../../exceptions'
import { createCommands } from './create-commands'

export const getTemplate = async (
  context: ServiceContext,
  id: string,
  version?: number
): Promise<TemplateVersion> => {
  if (!id) {
    throw MissingAttributes(['id'])
  }
  const db = ensureClient<IDBClient>(context, 'db')
  const params = createCommands(id, version, db.tables.templates.name)
  const dbItem = await db.get<DBTemplateItem>(params)

  if (dbItem) {
    return removeKeyInfo(dbItem)
  }

  throw TemplateNotFound(id)
}
