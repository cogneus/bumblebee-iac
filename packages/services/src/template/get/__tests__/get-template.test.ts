import { MissingAttributes, TemplateNotFound } from '../../../exceptions'
import { serviceContext } from '../../__mocks__/service-context.mock'
import { dbTemplateItem } from '../../__mocks__/template.mock'
import { getTemplate } from '../get-template'
import { NIL as emptyId } from 'uuid'

describe('Get template tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  test('latest template request should match snapshot', async () => {
    let input = null

    ;(serviceContext.clients.db?.get as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return dbTemplateItem
    })
    const result = await getTemplate(serviceContext, emptyId)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })

  test('template version request should match snapshot', async () => {
    let input = null
    ;(serviceContext.clients.db?.get as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return dbTemplateItem
    })
    const result = await getTemplate(serviceContext, emptyId, 1)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })

  test('request should throw error if id is null', async () => {
    expect(async () => {
      await getTemplate(serviceContext, null as any)
    }).rejects.toThrow(MissingAttributes(['id']))
  })

  test('request should throw error if not found', async () => {
    ;(serviceContext.clients.db?.get as jest.Mock).mockImplementationOnce(() => undefined)
    expect(async () => {
      await getTemplate(serviceContext, emptyId)
    }).rejects.toThrow(TemplateNotFound(emptyId))
  })
})
