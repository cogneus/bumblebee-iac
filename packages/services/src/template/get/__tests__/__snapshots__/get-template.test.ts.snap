// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get template tests latest template request should match snapshot 1`] = `
{
  "input": {
    "Key": {
      "PK": "template#00000000-0000-0000-0000-000000000000",
      "SK": "current",
    },
    "TableName": "name-templates",
  },
  "result": {
    "created": "2023-01-01T00:00:00Z",
    "group": "mock-group",
    "id": "00000000-0000-0000-0000-000000000000",
    "modified": "2023-01-01T00:00:00Z",
    "source": "mock-source",
    "target": "mock-target",
    "template": {},
    "version": 1,
  },
}
`;

exports[`Get template tests template version request should match snapshot 1`] = `
{
  "input": {
    "Key": {
      "PK": "template#00000000-0000-0000-0000-000000000000",
      "SK": "version#0000000001",
    },
    "TableName": "name-templates",
  },
  "result": {
    "created": "2023-01-01T00:00:00Z",
    "group": "mock-group",
    "id": "00000000-0000-0000-0000-000000000000",
    "modified": "2023-01-01T00:00:00Z",
    "source": "mock-source",
    "target": "mock-target",
    "template": {},
    "version": 1,
  },
}
`;
